### PR TITLE
feat(behavior_path_planner): update path when object is gone

### LINKED ISSUE #2272: `obstacle_stop_planner` does not revert its avoidance path after the avoided obstacle is gone.
### Checklist

- [X] I've read the [contribution guidelines](https://github.com/autowarefoundation/autoware/blob/main/CONTRIBUTING.md).
- [X] I've searched other issues and no duplicate issues were found.
- [X] I'm convinced that this is not my fault but a bug.

### Description

I have deployed several vehicles near a train station. It is very common for other vehicles to come, stop for a minute to drop off someone, then go away.

Although the avoidance works nicely while the vehicle is parked on the side of the road, autoware still wants to follow the avoidance path after the vehicle is gone. This is very akward for the safety driver, as autoware moves away from its lane for no apparent reason

This can be reproduced with the planning simulator easily:

1. No obstacle -> no avoidance
Screenshot from 2022-11-12 08-42-38

2. A vehicle comes and park -> avoidance path is generated after a few seconds
Screenshot from 2022-11-12 08-43-34

3. The vehicle is gone, but autoware does not revert the path to 1 (even if you wait for a very long time)
Screenshot from 2022-11-12 08-44-05

### Expected behavior

If there is no obstacle, there should be no avoidance

### Actual behavior

generated avoidance path stays for ever.

### Steps to reproduce

See the pictures. You can use the fake bus to emulate a parked vehicle, and delete all obstacles after avoidance path is generated. 

### Versions

_No response_

### Possible causes

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ismetatabay**
Signed-off-by: ismet atabay <ismet@leodrive.ai>

## Description

At avoidance module in behavior_path_planner, if the detected objects occur in path,  avoidance module shift path points to around the vehicle. Moreover, if the detected objects are gone the shifted path is not updated. The reason of this, avoidance module' current state stays in the "RUNNING" state until the vehicle reach its goal. I added a feature to change this state, and this feature allows the path to be updated by completing the task of the avoidance module if the detected objects are gone.

To control this, I added a parameter with the name ```enable_update_path_when_object_is_gone``` and set the default value to ```false```. ```enable_update_path_when_object_is_gone``` must be ```true``` for this feature to be used.

## Related links

closes #2272

### Before PR (```enable_update_path_when_object_is_gone``` is set to ```false```):

https://user-images.githubusercontent.com/56237550/202670238-1a310bef-5404-4c9e-95cf-10283cc2cdb6.mp4

#### Case 1: The vehicle stops.

https://user-images.githubusercontent.com/56237550/203820169-6886c14b-dd3d-49e3-9396-319ca0da52b9.mp4

#### Case 2: The vehicle moves, avoidance maneuver doesn't starts.

https://user-images.githubusercontent.com/56237550/203820408-5f751017-c268-4bd0-a541-ae70b5118c76.mp4

#### Case 3: The vehicle moves, avoidance maneuver starts.

https://user-images.githubusercontent.com/56237550/203820615-4a5c3674-bcf9-4555-8757-88ecf57bd33a.mp4

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by satoshi-ota** [comment]
@ismetatabay Thanks for your contribution :smile: 

Could you show the ego behavior in case where the ego is in avoidance maneuver? I think it dangerous that the path suddenly reverts back to its original one when there is huge lateral deviation between the ego and reverted path.

**Comment by ismetatabay** [comment]
@satoshi-ota Hi, 
Thank you for comment, the behavior of the car is as in the video below. What do you suggest in this situation? Maybe we can revert the path if the vehicle is stopped, but we do not change the path while the vehicle is in motion.

https://user-images.githubusercontent.com/56237550/202898463-19d1d01e-bf21-44e6-81ae-06b5db5bcec2.mp4

**Comment by satoshi-ota** [comment]
@ismetatabay Thanks for your explanation :+1: 

The following scene is exactly the scenario I was concerned about. In this case, I think it would be better to replan the return path from Ego's position back to the original reference path, because it is an abnormal situation for a system where the vehicle deviates significantly from its path, and I think we have to avoid such a situation as much as possible. What would you think?

As another idea, I think it would be good to make the assumption that the canceling of the avoidance maneuver is limited to only when the ego has not initiated the avoidance and there is not much deviation between the ego and reverted path.

image

**Comment by ismetatabay** [comment]
@satoshi-ota Thanks also for suggesting 2 solutions, 
With these solutions we can get a reasonable behavior. In terms of safety, your second suggestion seemed safer to me. An algorithm like this looks good to me:

If the vehicle has not started the avoidance maneuver, the avoidance path can be reverted, if the avoidance maneuver has been started, remain the avoidance path to follow it. The avoidance path follows a very smooth path to return to the old path. So it made sense to me to follow this avoid path to avoid sudden lateral change.

If that's okay I can update the algorithm.

**Comment by satoshi-ota** [comment]
@ismetatabay 

> In terms of safety, your second suggestion seemed safer to me.

I think so too :+1: 
Additionaly, I also think this method is easier to implement than the another one.

> If that's okay I can update the algorithm.

Yes, I appreciate your help so much.

**Comment by ismetatabay** [comment]
@satoshi-ota Thanks again for your suggestions, I'm a little busy for the next few days, then I'll update the algorithm 👍.

**Comment by ismetatabay** [comment]
@satoshi-ota hi, I updated the algorithm and the videos in the description according to the cases we talked about earlier :+1: .

**Comment by satoshi-ota** [comment]
@ismetatabay Thanks a lot :+1: I'll review this, so please wait a few days.

**Comment by ismetatabay** [comment]
@satoshi-ota hi, I added "Avoidance Canceling" section to the documentation. At this section, I explain Avoidance Module behavior at each different cases.

**Review by satoshi-ota — APPROVED** [review]
@ismetatabay Perfect :+1:
Thanks for your contribution!

File: planning/behavior_path_planner/behavior_path_planner_avoidance-design.md [hunk 1]
```diff
@@ -380,6 +380,16 @@ The shift points are modified by a filtering process in order to get the expecte
 - Similar gradient removal: Connect two shift points with a straight line, and remove the shift points in between if their shift amount is in the vicinity of the straight line.
 - Remove momentary returns: For shift points that reduce the avoidance width (for going back to the center line), if there is enough long distance in the longitudinal direction, remove them.
 
+##### Avoidance Cancelling
+
+If `enable_update_path_when_object_is_gone` parameter is true, Avoidance Module takes different actions according to the situations as follows:
+
+- If vehicle stops: If there is any object in the path of the vehicle, the avoidance path is generated. If this object goes away while the vehicle is stopping, the avoidance path will cancelled.
+- If vehcile is in motion, but avoidance maneuver doesn't started: If there is any object in the path of the vehicle, the avoidance path is generated. If this object goes away while the vehicle is not started avoidance maneuver, the avoidance path will cancelled.
```
Comment (satoshi-ota) [inline planning/behavior_path_planner/behavior_path_planner_avoidance-design.md:388]: @ismetatabay Sorry, could you fix this typo? :sweat_smile: 

vehcile -> vehicle

File: planning/behavior_path_planner/behavior_path_planner_avoidance-design.md [hunk 2]
```diff
@@ -380,6 +380,16 @@ The shift points are modified by a filtering process in order to get the expecte
...
+- If vehcile is in motion, avoidance maneuver started: If there is any object in the path of the vehicle, the avoidance path is generated,but if this object goes away while the vehicle is started avoidance maneuver, the avoidance path will not cancelled.
```
Comment (satoshi-ota) [inline planning/behavior_path_planner/behavior_path_planner_avoidance-design.md:389]: vehcile -> vehicle

File: planning/behavior_path_planner/behavior_path_planner_avoidance-design.md (hunk 1 above)
Comment (ismetatabay) [reply planning/behavior_path_planner/behavior_path_planner_avoidance-design.md]: @satoshi-ota Sorry i missed it, i updated it now. 👍

**Comment by satoshi-ota** [comment]
@ismetatabay Thanks :smile: You can merge this PR :tada:

**Pull request merged**